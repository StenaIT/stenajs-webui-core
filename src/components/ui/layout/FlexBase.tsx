import { CSSProperties } from 'react';
import * as React from 'react';

export type JustifyContentValue =
  | 'flex-start'
  | 'center'
  | 'flex-end'
  | 'space-around'
  | 'space-between';

export interface FlexBaseProps {
  flexDirection: 'row' | 'column';
}

export interface FlexBaseCommonProps {
  justifyContent?: JustifyContentValue;
  alignItems?: 'flex-start' | 'center' | 'flex-end';
  alignSelf?: 'flex-start' | 'center' | 'flex-end';
  minWidth?: string;
  flexGrow?: number;
  width?: string;
  height?: string;
  minHeight?: string;
  backgroundColor?: string;
  paddingVertical?: string;
  paddingHorizontal?: string;
  border?: string;
  className?: string;
  hidden?: boolean;
  style?: CSSProperties;
  tabIndex?: number;
}

export class FlexBase extends React.Component<
  FlexBaseProps & FlexBaseCommonProps
> {
  render() {
    const {
      children,
      justifyContent,
      alignItems,
      alignSelf,
      flexGrow,
      width,
      minWidth,
      height,
      minHeight,
      backgroundColor,
      paddingVertical,
      paddingHorizontal,
      border,
      hidden,
      style,
      tabIndex,
      flexDirection,
    } = this.props;

    return (
      <div
        tabIndex={tabIndex}
        style={{
          display: hidden ? 'none' : 'flex',
          flexDirection,
          justifyContent,
          alignItems,
          alignSelf,
          flexGrow,
          width,
          minWidth: minWidth || width,
          height,
          minHeight,
          backgroundColor,
          border,
          paddingTop: paddingVertical,
          paddingBottom: paddingVertical,
          paddingLeft: paddingHorizontal,
          paddingRight: paddingHorizontal,
          ...style,
        }}
      >
        {children}
      </div>
    );
  }
}
