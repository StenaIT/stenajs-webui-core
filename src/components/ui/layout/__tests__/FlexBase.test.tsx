import { shallow } from 'enzyme';
import * as React from 'react';
import { FlexBase, FlexBaseProps } from '../FlexBase';

describe('FlexBase', () => {
  const props: FlexBaseProps = {
    flexDirection: 'row',
  };

  it('must not change', () => {
    const wrapper = shallow(
      <FlexBase
        {...props}
        justifyContent={'center'}
        alignItems={'flex-end'}
        alignSelf={'flex-start'}
        flexGrow={1}
        width={'30px'}
        minWidth={'45px'}
        height={'60px'}
        minHeight={'50px'}
        backgroundColor={'#fff'}
        paddingVertical={'10px'}
        paddingHorizontal={'15px'}
        border={'1px solid black'}
        hidden
        tabIndex={0}
        style={{ color: '#000' }}
      />,
    );
    expect(wrapper).toMatchSnapshot();
  });

  describe('hidden', () => {
    describe('when hidden is provided', () => {
      it("sets display to 'none'", () => {
        const wrapper = shallow(<FlexBase hidden {...props} />);
        expect(wrapper.find('div').prop('style')!.display).toBe('none');
      });
    });
    describe('when hidden is not provided', () => {
      it("sets display to 'flex'", () => {
        const wrapper = shallow(<FlexBase {...props} />);
        expect(wrapper.find('div').prop('style')!.display).toBe('flex');
      });
    });
  });

  describe('minWidth', () => {
    describe('when minWidth is provided', () => {
      it('sets minWidth to minWidth', () => {
        const minWidth = '30px';
        const wrapper = shallow(<FlexBase minWidth={minWidth} {...props} />);
        expect(wrapper.find('div').prop('style')!.minWidth).toBe(minWidth);
      });
    });
    describe('when minWidth is not provided', () => {
      it('sets minWidth to width', () => {
        const width = '30px';
        const wrapper = shallow(<FlexBase width={width} {...props} />);
        expect(wrapper.find('div').prop('style')!.minWidth).toBe(width);
      });
    });
  });
});
