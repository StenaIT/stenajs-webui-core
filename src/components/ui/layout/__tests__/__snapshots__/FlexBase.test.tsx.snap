// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlexBase must not change 1`] = `
<div
  style={
    Object {
      "alignItems": "flex-end",
      "alignSelf": "flex-start",
      "backgroundColor": "#fff",
      "border": "1px solid black",
      "color": "#000",
      "display": "none",
      "flexDirection": "row",
      "flexGrow": 1,
      "height": "60px",
      "justifyContent": "center",
      "minHeight": "50px",
      "minWidth": "45px",
      "paddingBottom": "10px",
      "paddingLeft": "15px",
      "paddingRight": "15px",
      "paddingTop": "10px",
      "width": "30px",
    }
  }
  tabIndex={0}
/>
`;
